React native - Use Core

igor-dv opened the pull request:
Issue: #4879 #4011 #4852 #4332 #5249 #5309

RN was a bit behind for a long time with all the new features we did. In order to support all of the good things, we should reuse as much as possible from the core. Now when the manager and preview are separated it's much easier to do.

@Gongreg, we can maybe continue the #4752 here or create a new branch from this one.
Issues we can also finish for the breaking v5 release:
- [x] #4879 
- [x] #4011 ~= #4852
- [x] #4332


## What I did
- moved `src/manager` and `src/preview` to `src/client/*` so it will be similar to other apps
- deleted all the webpack/babel/config related code from rn/server and replaced it with the `dev-server` from core
- ~changed `dev-server` to be able to run the webpack only for manager~
- used `standalone` api to build storybook in a dev mode
- removed `--secured` param (`-s`) and used `--https` like in other apps


- removed build command (it was not doing anything).

TBD:
- [ ] Validate breaking changes
- [x] Check something very useful is not missing now
- [ ] Check what to do now with the `environment` parameter
- [ ] Reflect everything in docs
- [ ] Update migration tutorial

igor-dv committed:
Use core/devServer in RN

igor-dv committed:
Delete unneeded code

igor-dv committed:
Move all the client files to client dir like in other apps

igor-dv committed:
Remove unneeded deps from RN

igor-dv committed:
Add rn-options-preset.js preset to pass options into manager

igor-dv committed:
Extend manager-preset to use custom managerEntry + use the original manager-preset in rn

igor-dv committed:
Merge remote-tracking branch 'origin/next' into react-native/use-core-for-server

igor-dv committed:
Simplify dev-server

igor-dv committed:
Update deps in the crna app

igor-dv commented:
Currently, when running an app with expo I am getting the babe6/babel7 mismatch...

igor-dv commented:
upgrading to sdk31 looks solving the babel issue, but there is something new now =/

igor-dv committed:
Upgrade sdk to 31 to solve babel7 issue + downgrade react to 16.5.1 to solve 'schedule/tracking' issue

igor-dv committed:
Update sdk in app.json

igor-dv committed:
Merge remote-tracking branch 'origin/next' into react-native/use-core-for-server

# Conflicts:
#	app/react-native/package.json

igor-dv committed:
Use `standalone` to build RN manager instead of just dev-server

ndelangen commented:
@Gongreg @igor-dv should we perhaps merge this efforts into the ui-overhaul?

This would be a new major release right?

igor-dv commented:
Tbh I prefer to have a separate branch for v5 that works. And merge UI overhaul there once it is ready. Otherwise we won't be able to relaease alphas to check things

Gongreg committed:
Set storybook folder to `.storybook`.

Gongreg commented:
@igor-dv, solved #4332

Gongreg committed:
Removed build command.

Gongreg committed:
Merge next

Gongreg committed:
Removed config.

Gongreg committed:
@storybook/react-native dependency now works.

Gongreg committed:
renamed storybook to .storybook in react-native example

igor-dv committed:
Merge branch 'next' into react-native/use-core-for-server

Gongreg committed:
moving react to peer dependency.

igor-dv committed:
Delete conflict files

igor-dv committed:
Set eventemitter3 to deps in client-api

igor-dv committed:
Fix rn generators linter

igor-dv committed:
Add missing deps to rn-server

igor-dv committed:
Fix linter one more time

Gongreg committed:
Merge remote-tracking branch 'origin/react-native/use-core-for-server' into react-native/use-core-for-server

igor-dv committed:
storyshots should also relate to the .storybook dir

Gongreg committed:
Merge remote-tracking branch 'origin/react-native/use-core-for-server' into react-native/use-core-for-server

Gongreg committed:
Asks if you want to install @storybook/react-native-server

igor-dv committed:
Not all props from options were passed to the provider

Gongreg committed:
Reverting back to storybook.

Gongreg committed:
Reverting back to storybook.

Gongreg committed:
Creates http server, so websockets can connect to it.

maxkomarychev commented:
hey @igor-dv thanks for this, I'd love to play with your fixes asap even if they aren't merged with the upstream, how I can try this? I tried to using git url in my package json against your branch but it doesn't work, assuming this is because it's monorepo managed with lerna.

igor-dv commented:
Currently, the branch is broken. We need to fix a few things.

igor-dv committed:
Revert few more .storybook changes

igor-dv committed:
Merge branch 'next' into react-native/use-core-for-server

# Conflicts:
#	app/react-native/package.json
#	lib/addons/package.json
#	lib/addons/src/index.js
#	lib/addons/src/make-decorator.test.ts
#	lib/core/package.json

rkdavidson commented:
Hey @igor-dv , any clarity/direction on the things that need fixed so this can be merged?

Gongreg commented:
Hey. It should more or less work. We need to update documentation and tests things out mostly

Gongreg committed:
Updated documentation.

Gongreg committed:
Updated server documentation.

Gongreg committed:
updated argument

Gongreg commented:
Hey @igor-dv,
It is needed for react-native-client. But we can do it seperately. :)

igor-dv commented:
@Gongreg , cool, I think let's do anything we can do separably, it will simplify the merge

igor-dv committed:
Merge remote-tracking branch 'origin/next' into react-native/use-core-for-server

# Conflicts:
#	addons/ondevice-knobs/package.json
#	app/react-native/package.json
#	app/react-native/readme.md
#	examples-native/crna-kitchen-sink/package.json
#	lib/addons/package.json
#	lib/addons/src/make-decorator.test.ts
#	lib/client-api/package.json
#	lib/client-api/src/client_api.test.js
#	lib/client-api/src/index.js
#	lib/client-api/src/story_store.js
#	lib/client-api/src/story_store.test.js
#	lib/core/package.json
#	lib/core/src/client/preview/index.js
#	lib/core/src/client/preview/start.js
#	lib/core/src/server/build-dev.js
#	lib/core/src/server/dev-server.js
#	lib/core/src/server/manager/manager-preset.js
#	yarn.lock

igor-dv committed:
Bring back files deleted in merge

igor-dv committed:
Fix linter

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

# Conflicts:
#	app/react-native/package.json
#	yarn.lock

ndelangen committed:
FIX tests a bit

ndelangen commented:
This seems to be an issue:
```
Running smoke test in react_native_scripts
$ storybook start --smoke-test --quiet
info => Loading presets
info => Loading custom addons config.


WARN Unable to close preview build!


WARN FATAL broken build!, will close the process,
WARN Fix the error below and restart storybook.

ModuleBuildError: Module build failed (from /Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js):
Error: Plugin/Preset files are not allowed to export objects, only functions. In /Users/dev/Projects/GitHub/storybook/core/node_modules/babel-preset-expo/index.js
    at createDescriptor (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:178:11)
    at items.map (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:109:50)
    at Array.map (<anonymous>)
    at createDescriptors (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:109:29)
    at createPresetDescriptors (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:101:10)
    at presets (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:47:19)
    at mergeChainOpts (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-chain.js:320:26)
    at /Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-chain.js:283:7
    at buildRootChain (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-chain.js:120:22)
    at loadPrivatePartialConfig (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/partial.js:85:55)
    at Object.loadPartialConfig (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/partial.js:110:18)
    at Object.<anonymous> (/Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:140:26)
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:3:103)
    at _next (/Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:5:194)
    at /Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:5:364
    at runLoaders (/Users/dev/Projects/GitHub/storybook/core/node_modules/webpack/lib/NormalModule.js:301:20)
    at /Users/dev/Projects/GitHub/storybook/core/node_modules/loader-runner/lib/LoaderRunner.js:367:11
    at /Users/dev/Projects/GitHub/storybook/core/node_modules/loader-runner/lib/LoaderRunner.js:233:18
    at context.callback (/Users/dev/Projects/GitHub/storybook/core/node_modules/loader-runner/lib/LoaderRunner.js:111:13)
    at loader.call.then.err (/Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:51:103)
    at <anonymous>
    at runMicrotasksCallback (internal/process/next_tick.js:121:5)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
##teamcity[testFailed name='react_native_scripts']
##teamcity[testFinished name='react_native_scripts']
##teamcity[testStarted name='react_project' captureStandardOutput='true']
```

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

ndelangen committed:
ADD --yes option to CLI init command && USE --yes command to skip question where possible

ndelangen committed:
FIX the --yes switch for RN

ndelangen committed:
FIX an issue with storyshots-puppeteer tests.

If you have font rendering inconsistencies, you just remove all text like a pro

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

# Conflicts:
#	lib/node-logger/src/index.test.ts

ndelangen commented:
I fixed the issue of the cli test being stuck on running init

Other issue of the smoke-test remains

ndelangen committed:
ADD prepare to react-native package

ndelangen commented:
@igor-dv It kinda looks like we need to upgrade to babel7 for Rn to make this work?

I'm not sure how it passed before?

Maybe it's the change we did here that is surfacing an existing issue?

igor-dv commented:
Babel 7 was a part of these changes

ndelangen committed:
WIP

ndelangen committed:
WIP

gorhgorh committed:
WIP

shilman committed:
WIP: Set up RN story store in getStorybookUI

tmeasday commented:
Going to remove this from v5 now the decision has been made to release separately

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

# Conflicts:
#	app/react-native/package.json
#	yarn.lock

ndelangen committed:
BRING the reactnative example into the workspace

benoitdion committed:
Restore crna-kitchen-sink

Unfortunately expo is stuck with an older version of react. We can't be in the yarn workspace just yet.

benoitdion committed:
Disable preview for now to avoid crashing the server

benoitdion committed:
Give parameters defaults

benoitdion committed:
Fix app/react-native

- Bind fns in constructors and initialize StoryStore
- Make channel optional, we might not have one yet.
- Use getElements instead of getPanels

ndelangen committed:
CHANGE to run server from workspace, without the example being in the workspace

benoitdion committed:
Fix event names

benoitdion committed:
Disable preview since it hasn't been implemented yet

benoitdion committed:
Some documentation

benoitdion committed:
Merge remote-tracking branch 'upstream/react-native/use-core-for-server' into react-native/use-core-for-server

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

# Conflicts:
#	addons/actions/package.json
#	addons/cssresources/package.json
#	lib/core/package.json
#	lib/ui/src/keyboard/keyCodes.ts
#	yarn.lock

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

# Conflicts:
#	addons/a11y/package.json
#	addons/actions/package.json
#	addons/events/package.json
#	addons/info/package.json
#	addons/jest/package.json
#	addons/knobs/package.json
#	addons/links/package.json
#	addons/ondevice-knobs/package.json
#	addons/ondevice-notes/package.json
#	app/react-native/package.json
#	app/react/package.json
#	examples-native/crna-kitchen-sink/package.json
#	examples/cra-kitchen-sink/package.json
#	examples/official-storybook/package.json
#	lib/components/package.json
#	lib/theming/package.json
#	lib/ui/package.json
#	package.json
#	yarn.lock

shilman commented on addons/ondevice-knobs/package.json:
```diff
@@ -1,7 +1,8 @@
-  "description": "Display storybook story knobs on your device.",
+  "private": true,
+  "version": "5.0.0",
+  "description": "Display storybook story knobs on your deviced.",
```
deviced typo

shilman reviewed (approved):
LGTM!

benoitdion committed:
Merge remote-tracking branch 'upstream/next' into react-native/use-core-for-server

benoitdion committed:
update versions to 5.1.0-alpha.4

benoitdion committed:
fix android host

benoitdion committed:
Fix server story list rendering

shilman commented on addons/ondevice-knobs/src/panel.js:
```diff
@@ -1,7 +1,8 @@
 import PropTypes from 'prop-types';
-
+import { SELECT_STORY, FORCE_RE_RENDER } from '@storybook/core-events';
+import { SET, SET_OPTIONS, RESET, CHANGE, CLICK } from '@storybook/addon-knobs/dist/shared';
```
We shouldn't be importing from `dist` -- let's export this in `@storybook/addon-knobs/events`?

shilman commented on app/react-native-server/readme.md:
```diff
@@ -0,0 +1,13 @@
+# Storybook Server for React Native
+Additional package for @storybook/react-native to support server.
```
=> `to support a web server and browser UI that runs alongside the RN on-device UI`

shilman commented on app/react-native/package.json:
```diff
@@ -18,64 +18,28 @@
   },
   "license": "MIT",
-  "main": "dist/index.js",
+  "main": "src/index.js",
```
Isn't this the non-transpiled version?

shilman commented on app/react-native/readme.md:
```diff
@@ -90,7 +82,7 @@ The following parameters can be passed to the start command:
 -p, --port <port>
     port to listen on
--s, --secured
+--https
```
Needs a note in `MIGRATION.md` and/or should be deprecated with one major version of back-compat

ndelangen replied:
I will add it to the MIGRATION.MD

shilman commented on app/react-native/src/preview/components/OnDeviceUI/index.js:
```diff
@@ -3,11 +3,11 @@ import PropTypes from 'prop-types';
-import StoryListView from '../StoryListView';
-import StoryView from '../StoryView';
-import Addons from './addons';
+import StoryListView from '../StoryListView/index';
```
Isn't `index` implicit?

ndelangen replied:
Yes, but I don't recommend relying on that behavior

shilman replied:
Why not? Doesn't the majority of JS code (including most of the storybook codebase) rely on that?

benoitdion replied:
removed the `/index` since it is implicit and is consistent with the rest of the codebase

shilman commented on app/react-native/src/preview/index.js:
```diff
@@ -1,24 +1,25 @@
-import OnDeviceUI from './components/OnDeviceUI';
-import StoryView from './components/StoryView';
+import { StoryStore, ClientApi } from '@storybook/client-api';
+import OnDeviceUI from './components/OnDeviceUI/index';
```
index implicit

ndelangen replied:
Right, but makes it harder to find, if neglected.

shilman commented on app/react-native/src/preview/index.js:
```diff
@@ -132,10 +134,19 @@ export default class Preview {
+    if (params.host)
+      return params.host;
+    if (Platform.OS === 'android')
+      return '10.0.2.2';
```
Is this documented somewhere?

shilman commented on examples-native/crna-kitchen-sink/package.json:
```diff
@@ -34,18 +36,22 @@
     "@storybook/core-events": "file:../../packs/storybook-core-events.tgz",
     "@storybook/node-logger": "file:../../packs/storybook-node-logger.tgz",
     "@storybook/react-native": "file:../../packs/storybook-react-native.tgz",
+    "@storybook/react-native-server": "file:../../app/react-native-server",
```
Why not `.tgz`? Can we remove the other `.tgz` files and remove the `build-packs` step entirely?

shilman commented on lib/cli/bin/generate.js:
```diff
@@ -22,6 +22,7 @@ if (process.argv[1].includes('getstorybook')) {
     .option('-N --use-npm', 'Use npm to install deps')
     .option('-p --parser <babel | babylon | flow>', 'jscodeshift parser')
     .option('-t --type <type>', 'Add Storybook for a specific project type')
+    .option('-y --yes', 'Answer yes to all prompts')
```
Should this be a separate PR?

shilman reviewed (changes requested):
So happy this is almost mergeable. Some minor nitpicks above, but otherwise it looks great to me.

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

# Conflicts:
#	addons/ondevice-knobs/package.json
#	app/react-native/package.json

ndelangen committed:
FIX missing .tgz

ndelangen committed:
ADD note to MIGRATION.MD about the change to the start command for https for RN

ndelangen committed:
IMRPOVE the readme description for the RN server package

ndelangen committed:
CHANGE the react-native package to use the dist

ndelangen committed:
ADD react-native-server to cli fixtures tests

ndelangen committed:
Merge branch 'next' into react-native/use-core-for-server

# Conflicts:
#	addons/ondevice-knobs/package.json
#	app/react-native/package.json

ndelangen committed:
UPGRADE the react-native-scripts to latest expo && ADD app/react-native as dependency to RN projects (always)

ndelangen committed:
FIX location of tgz

benoitdion committed:
Use rn-host-detect to identify host

benoitdion committed:
Clean up imports

benoitdion committed:
Merge branch 'next' into react-native/use-core-for-server

shilman reviewed (approved):
Awesome!!!

ndelangen committed:
FIX docs yarn.lock

igor-dv commented:
Oops I am late. Yay !! Great Success 💥

dangreenisrael commented:
🎉
